bundles/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/internal/NeeoRoomProtocol.java

```diff
@@ -123,19 +123,23 @@ public void processAction(NeeoAction action) {
     private void processScenarioChange(String scenarioKey, boolean launch) {
         NeeoUtil.requireNotEmpty(scenarioKey, "scenarioKey cannot be empty");
 
-        final String[] activeScenarios = this.activeScenarios.get();
-        final int idx = ArrayUtils.indexOf(activeScenarios, scenarioKey);
+        final List<String> activeScenarios = this.activeScenarios.get();
 
-        // already set that way
-        if ((idx < 0 && !launch) || (idx >= 0 && launch)) {
-            return;
+        if (activeScenarios.contains(scenarioKey)) {
+            if (launch) {
+                return;
+            } else {
+                activeScenarios.remove(scenarioKey);
+            }
+        } else {
+            if (launch) {
+                activeScenarios.add(scenarioKey);
+            } else {
+                return;
+            }
         }
 
-        final String[] newScenarios = idx >= 0 ? (String[]) ArrayUtils.remove(activeScenarios, idx)
-                : (String[]) ArrayUtils.add(activeScenarios, scenarioKey);
-
-        this.activeScenarios.set(newScenarios);
-
+        this.activeScenarios.set(activeScenarios);
```

Comment:
For this to behave correctly in a concurrent environment you should use `AtomicReference.compareAndSet` to make sure that the reference has't been changed by something else before you were able to set it. Usually this is put in a while loop and only exists once the compareAndSet returns true.

Reply:
I wonder if this is sort of over-kill, but I changed it accordingly.

Reply:
I'm sure it is, but if any of the atomic utility classes are going to be used, they should be used correctly.